s3/s3.go

```diff
@@ -799,7 +802,7 @@ func (f *Fs) Copy(src fs.Object, remote string) (fs.Object, error) {
 
 // Hashes returns the supported hash sets.
 func (f *Fs) Hashes() fs.HashSet {
-	return fs.HashSet(fs.HashMD5)
```

Comment:
Yes, I thought about getting the best of both ETag and MD5.  The issue is that, for each file, either MD5 or ETag will be optimal, but we commit to a hash algorithm at the filesystem level.  It's even a bit worse because an algorithm is chosen for a filesystem without regard to any criteria.  So, I wanted to support the one hash algorithm that is the best in most cases.  Based on the chart below, I chose ETag over MD5.  What do you think?

Scenario|MD5|ETag
---|---|---
Small file, rcloned to S3|X|X
Large file, rcloned to S3| |X
Small file, 3rd-party put to S3|X|X
Large file, 3rd-party put to S3| |
KMS/client encrypted file| |
Small file, rclone checked across remotes|X|
Large file, rclone checked across remotes| |

Reply:
As an alternative here, the hash handshaking could be enhanced to support file-by-file negotiation; parameterized algorithms, discovering algorithm parameters during the handshake, etc...

I'd be interested in contributing to enhanced hash handshaking; let me know if you'd like to chat!

Reply:
There is one thing that worries me about your proposal is that this would make rclone incompatible with other methods of transferring data.  So say someone has a workflow of using cyberduck to upload stuff and rclone to retrieve it remotely, then that might break with corrupted hash errors.  rclone is in use by 1,000s of people so I don't want to take that risk.

So I think your proposal as it stands needs to be behind an `--s3-etag` flag - if set it would use s3etags, if not then md5sums and blank for the multipart uploads.

However I think we could do better than the PartSize function... given an etag with hash+partitions and the files length, and given that all the parts are the same size (except for the last), then it is possible to uniquely determine the correct partition size to use (I managed to convince myself of this).  So we should be able to check any etag which was uploaded with constant size parts. I'm unsure how we'd fit this into rclone though, so maybe what you've done behind a flag is the best of both worlds.

> I'd be interested in contributing to enhanced hash handshaking; let me know if you'd like to chat!

Your ideas would be appreciated!